<h1>Final Capstone Project</h1>

### This notebook is on Segmenting and Clustering Neighborhoods in Toronto - Week 3 of Capstone



# Table of Contents 

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1.  <a href="#item1">Week 3 - Section 1</a>

2.  <a href="#item2">Week 3: Section 2</a>

3.  <a href="#item3">Week 3: Section 3</a>

    </font>
    </div>

In [1]:
import pandas as pd
import numpy as np
print ("Assignments for Week 3 -- all in 1 Notebook")

Assignments for Week 3 -- all in 1 Notebook


## Week 3 - Section 1:
### Scraping from website, examining data, and pulling into dataframe:

In [2]:
import requests

!pip install lxml

!pip install html5lib

!pip install bs4
from bs4 import BeautifulSoup

#!pip install folium
!pip install folium==0.5.0


In [3]:
website = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
r = requests.get(website)
r

<Response [200]>

In [4]:
# examine the data
html=r.content
soup = BeautifulSoup(html, 'html.parser')
soup

<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"YBWHsZO6OLX47dOfGpgT8AAAAEM","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":995657573,"wgRevisionId":995657573,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description is different from Wikidata","Communications in Ontario","Postal co

In [5]:
#dfs = pd.read_html(website, header=0)
#dfs = pd.read_html(website, header=0, index_col=None)
#dfs = pd.read_html(website, header=0, index_col="Postal Code")
# Grabbing the table only
#CanPostal= dfs[0]
CanPostal = pd.read_html(website, header=0)[0]

In [6]:
CanPostal.columns

Index(['Postal Code', 'Borough', 'Neighbourhood'], dtype='object')

In [7]:
CanPostal.shape

(180, 3)

In [8]:
CanPostal.dtypes



Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

### Examining rows with not assigned, and removing rows with Boroughs that aren't assigned

In [9]:
CanPostalBNA = CanPostal[CanPostal['Borough'] == 'Not assigned']
CanPostalBNA

Postal Code       Borough Neighbourhood
0           M1A  Not assigned  Not assigned
1           M2A  Not assigned  Not assigned
7           M8A  Not assigned  Not assigned
10          M2B  Not assigned  Not assigned
15          M7B  Not assigned  Not assigned
16          M8B  Not assigned  Not assigned
19          M2C  Not assigned  Not assigned
24          M7C  Not assigned  Not assigned
25          M8C  Not assigned  Not assigned
28          M2E  Not assigned  Not assigned
29          M3E  Not assigned  Not assigned
33          M7E  Not assigned  Not assigned
34          M8E  Not assigned  Not assigned
35          M9E  Not assigned  Not assigned
37          M2G  Not assigned  Not assigned
38          M3G  Not assigned  Not assigned
42          M7G  Not assigned  Not assigned
43          M8G  Not assigned  Not assigned
44          M9G  Not assigned  Not assigned
51          M7H  Not assigned  Not assigned
52          M8H  Not assigned  Not assigned
53          M9H  Not assigned  Not assigned
60          M7J  Not assigned  Not assigned
61          M8J  Not assigned  Not assigned
62          M9J  Not assigned  Not assigned
69          M7K  Not assigned  Not assigned
70          M8K  Not assigned  Not assigned
71          M9K  Not assigned  Not assigned
78          M7L  Not assigned  Not assigned
79          M8L  Not assigned  Not assigned
..          ...           ...           ...
132         M7T  Not assigned  Not assigned
133         M8T  Not assigned  Not assigned
134         M9T  Not assigned  Not assigned
136         M2V  Not assigned  Not assigned
137         M3V  Not assigned  Not assigned
140         M6V  Not assigned  Not assigned
141         M7V  Not assigned  Not assigned
145         M2W  Not assigned  Not assigned
146         M3W  Not assigned  Not assigned
149         M6W  Not assigned  Not assigned
150         M7W  Not assigned  Not assigned
154         M2X  Not assigned  Not assigned
155         M3X  Not assigned  Not assigned
158         M6X  Not assigned  Not assigned
159         M7X  Not assigned  Not assigned
161         M9X  Not assigned  Not assigned
162         M1Y  Not assigned  Not assigned
163         M2Y  Not assigned  Not assigned
164         M3Y  Not assigned  Not assigned
166         M5Y  Not assigned  Not assigned
167         M6Y  Not assigned  Not assigned
170         M9Y  Not assigned  Not assigned
171         M1Z  Not assigned  Not assigned
172         M2Z  Not assigned  Not assigned
173         M3Z  Not assigned  Not assigned
174         M4Z  Not assigned  Not assigned
175         M5Z  Not assigned  Not assigned
176         M6Z  Not assigned  Not assigned
177         M7Z  Not assigned  Not assigned
179         M9Z  Not assigned  Not assigned

[77 rows x 3 columns]

In [10]:
CanPostal.drop(CanPostal[CanPostal.Borough == 'Not assigned'].index, inplace=True)
#CanPostal[~CanPostal.borough.str.contains("Not assigned")]
CanPostal.shape

(103, 3)

In [11]:
CanPostalNNA = CanPostal[CanPostal['Neighbourhood'] == 'Not assigned']
CanPostalNNA
# there are no neighbourhoods with "Not assigned" after removing others.

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

In [12]:
CanPostal = CanPostal.reset_index(drop=True)


In [13]:
CanPostal.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

## Output of data scraped is 179 total postal codes
## However, 77 rows have Borough of 'Not assigned'.  These were removed. In addition, no other rows have neighborhoods that aren't assigned so no action is needed on neighborhoods.  The final number of 103 postal codes in Toronto match the website figure at the top of the table.

In [14]:
CanPostal.shape
print("Final Shape: ",CanPostal.shape)

Final Shape:  (103, 3)


# Week 3:  Section 2

In [15]:
coordfile = 'http://cocl.us/Geospatial_data'
cf=pd.read_csv(coordfile,header=0,index_col='Postal Code')
cf.head(10)

Latitude  Longitude
Postal Code                      
M1B          43.806686 -79.194353
M1C          43.784535 -79.160497
M1E          43.763573 -79.188711
M1G          43.770992 -79.216917
M1H          43.773136 -79.239476
M1J          43.744734 -79.239476
M1K          43.727929 -79.262029
M1L          43.711112 -79.284577
M1M          43.716316 -79.239476
M1N          43.692657 -79.264848

In [16]:
cf.dtypes

Latitude     float64
Longitude    float64
dtype: object

In [17]:
cf.columns

Index(['Latitude', 'Longitude'], dtype='object')

In [18]:
#cf.join(CanPostal.set_index('Postal Code'), on='Postal Code')
j=CanPostal.join(cf.set_index(CanPostal['Postal Code']), on='Postal Code')
j

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
5           M9A         Etobicoke   
6           M1B       Scarborough   
7           M3B        North York   
8           M4B         East York   
9           M5B  Downtown Toronto   
10          M6B        North York   
11          M9B         Etobicoke   
12          M1C       Scarborough   
13          M3C        North York   
14          M4C         East York   
15          M5C  Downtown Toronto   
16          M6C              York   
17          M9C         Etobicoke   
18          M1E       Scarborough   
19          M4E      East Toronto   
20          M5E  Downtown Toronto   
21          M6E              York   
22          M1G       Scarborough   
23          M4G         East York   
24          M5G  Downtown Toronto   
25          M6G  Downtown Toronto   
26          M1H       Scarborough   
27          M2H        North York   
28          M3H        North York   
29          M4H         East York   
..          ...               ...   
73          M4R   Central Toronto   
74          M5R   Central Toronto   
75          M6R      West Toronto   
76          M7R       Mississauga   
77          M9R         Etobicoke   
78          M1S       Scarborough   
79          M4S   Central Toronto   
80          M5S  Downtown Toronto   
81          M6S      West Toronto   
82          M1T       Scarborough   
83          M4T   Central Toronto   
84          M5T  Downtown Toronto   
85          M1V       Scarborough   
86          M4V   Central Toronto   
87          M5V  Downtown Toronto   
88          M8V         Etobicoke   
89          M9V         Etobicoke   
90          M1W       Scarborough   
91          M4W  Downtown Toronto   
92          M5W  Downtown Toronto   
93          M8W         Etobicoke   
94          M9W         Etobicoke   
95          M1X       Scarborough   
96          M4X  Downtown Toronto   
97          M5X  Downtown Toronto   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.806686 -79.194353  
1                                     Victoria Village  43.784535 -79.160497  
2                            Regent Park, Harbourfront  43.763573 -79.188711  
3                     Lawrence Manor, Lawrence Heights  43.770992 -79.216917  
4          Queen's Park, Ontario Provincial Government  43.773136 -79.239476  
5              Islington Avenue, Humber Valley Village  43.744734 -79.239476  
6                                       Malvern, Rouge  43.727929 -79.262029  
7                                            Don Mills  43.711112 -79.284577  
8                      Parkview Hill, Woodbine Gardens  43.716316 -79.239476  
9                             Garden District, Ryerson  43.692657 -79.264848  
10                                           Glencairn  43.757410 -79.273304  
11   West Deane Park, Princess Gardens, Martin Grov...  43.750072 -79.295849  
12              Rouge Hill, Port Union, Highland Creek  43.794200 -79.262029  
13                                           Don Mills  43.781638 -79.304302  
14                                    Woodbine Heights  43.815252 -79.284577  
15                                      St. James Town  43.799525 -79.318389  
16                                  Humewood-Cedarvale  43.836125 -79.205636  
17   Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.803762 -79.363452  
18                   Guildwood, Morningside, West Hill  43.778517 -79.346556  
19                                         The Beaches  43.786947 -79.385975  
20                                         Berczy 

In [19]:
j.shape

(103, 5)

# Week 3:  Section 3

In [20]:
# Importing libraries:

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!pip install geopy
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')








Libraries imported.


### Identifying and mapping the neighbourhoods in Toronto

In [21]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="Toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

# create map of Toronto 
mapT = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighbourhood in zip(j['Latitude'], j['Longitude'], j['Borough'], j['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(mapT)  
    
mapT

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Defining Foursquare Credentials and Version

In [22]:
CLIENT_ID = '1CETT04VB4CLSAGO1IXLCHKKZHVZRSETNG0TXNWYPDA4SEY0' # your Foursquare ID
CLIENT_SECRET = 'Y2GSJIWDGC45VYYYGMV0ASLAC5FI21JECGRODK5GIRC31C0K' # your Foursquare Secret
ACCESS_TOKEN = 'RR1BSSE2OGRMAEYF3YDUYQ0AIDNGDXNM10VZ04ZZZ3DGLQYZ' # your FourSquare Access Token
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)
print('ACCESS_TOKEN:' + ACCESS_TOKEN)

Your credentails:
CLIENT_ID: 1CETT04VB4CLSAGO1IXLCHKKZHVZRSETNG0TXNWYPDA4SEY0
CLIENT_SECRET:Y2GSJIWDGC45VYYYGMV0ASLAC5FI21JECGRODK5GIRC31C0K
ACCESS_TOKEN:RR1BSSE2OGRMAEYF3YDUYQ0AIDNGDXNM10VZ04ZZZ3DGLQYZ


In [23]:
j.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

#### Looking at the first neighbourhood.



In [24]:
j.loc[0, 'Neighbourhood']

'Parkwoods'

In [25]:
neighborhood_latitude = j.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = j.loc[0, 'Longitude'] # neighborhood longitude value
neighborhood_name = j.loc[0, 'Neighbourhood'] # neighborhood name
print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                              neighborhood_longitude))

Latitude and longitude values of Parkwoods are 43.806686299999996, -79.19435340000001.


In [26]:
latitude = neighborhood_latitude 
longitude = neighborhood_longitude 
radius = 500
url = 'https://api.foursquare.com/v2/venues/search?client_id={}&client_secret={}&ll={},{}&oauth_token={}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude,ACCESS_TOKEN, VERSION, radius, LIMIT)
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=1CETT04VB4CLSAGO1IXLCHKKZHVZRSETNG0TXNWYPDA4SEY0&client_secret=Y2GSJIWDGC45VYYYGMV0ASLAC5FI21JECGRODK5GIRC31C0K&v=20180605&ll=43.806686299999996,-79.19435340000001&radius=500&limit=100'

In [27]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '6019b25bd7bbe43d3d90fcc0'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Malvern',
  'headerFullLocation': 'Malvern, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.8111863045, 'lng': -79.18812958073042},
   'sw': {'lat': 43.80218629549999, 'lng': -79.2005772192696}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bb6b9446edc76b0d771311c',
       'name': 'Wendy’s',
       'location': {'crossStreet': 'Morningside & Sheppard',
        'lat': 43.80744841934756,
        'lng': -79.19905558052072,
        'labeledLatLngs': [{'label': 'display'

#### Functions to loop through all venues in 1 neighborhood (Parkwood)

In [28]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [29]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()


name            categories        lat        lng
0                Wendy’s  Fast Food Restaurant  43.807448 -79.199056
1  Interprovincial Group            Print Shop  43.805630 -79.200378

In [30]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

2 venues were returned by Foursquare.


#### Functions to loop through all neighborhoods

In [31]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()['response']['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [32]:
Tor_venues = getNearbyVenues(names=j['Neighbourhood'], latitudes=j['Latitude'],longitudes=j['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [33]:

#Tor_venues.drop(Tor_venues.columns['Neighborhood'], axis=1, )
#Tor_venues.drop(['Neighborhood'], axis=1, inplace=True)
Tor_venues.head()


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                  Parkwoods              43.806686              -79.194353   
1                  Parkwoods              43.806686              -79.194353   
2           Victoria Village              43.784535              -79.160497   
3           Victoria Village              43.784535              -79.160497   
4  Regent Park, Harbourfront              43.763573              -79.188711   

                                               Venue  Venue Latitude  \
0                                            Wendy’s       43.807448   
1                              Interprovincial Group       43.805630   
2                              Royal Canadian Legion       43.782533   
3  SEBS Engineering Inc. (Sustainable Energy and ...       43.782371   
4                                     RBC Royal Bank       43.766790   

   Venue Longitude              Venue Category  
0       -79.199056        Fast Food Restaurant  
1       -79.200378                  Print Shop  
2       -79.163085                         Bar  
3       -79.156820  Construction & Landscaping  
4       -79.191151                        Bank

In [34]:
print(Tor_venues.shape)

(2117, 7)


In [35]:
Tor_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                  Parkwoods              43.806686              -79.194353   
1                  Parkwoods              43.806686              -79.194353   
2           Victoria Village              43.784535              -79.160497   
3           Victoria Village              43.784535              -79.160497   
4  Regent Park, Harbourfront              43.763573              -79.188711   

                                               Venue  Venue Latitude  \
0                                            Wendy’s       43.807448   
1                              Interprovincial Group       43.805630   
2                              Royal Canadian Legion       43.782533   
3  SEBS Engineering Inc. (Sustainable Energy and ...       43.782371   
4                                     RBC Royal Bank       43.766790   

   Venue Longitude              Venue Category  
0       -79.199056        Fast Food Restaurant  
1       -79.200378                  Print Shop  
2       -79.163085                         Bar  
3       -79.156820  Construction & Landscaping  
4       -79.191151                        Bank

In [36]:
print('There are {} uniques categories.'.format(len(Tor_venues['Venue Category'].unique())))

There are 269 uniques categories.


In [37]:
new_df = Tor_venues[Tor_venues['Venue Category'].str.contains('Neighborhood', na=False)]
new_df.head()

Neighborhood  Neighborhood Latitude  \
306                       Little Portugal, Trinity              43.676357   
434   Brockton, Parkdale Village, Exhibition Place              43.659526   
1021                   Birch Cliff, Cliffside West              43.650571   
1105                   Willowdale, Willowdale East              43.640816   

      Neighborhood Longitude             Venue  Venue Latitude  \
306               -79.293031     Upper Beaches       43.680563   
434               -79.340923       Leslieville       43.662070   
1021              -79.384568  Downtown Toronto       43.653232   
1105              -79.381752      Harbourfront       43.639526   

      Venue Longitude Venue Category  
306        -79.292869   Neighborhood  
434        -79.337856   Neighborhood  
1021       -79.385296   Neighborhood  
1105       -79.380688   Neighborhood

In [38]:
Tor_venues['Venue Category']= Tor_venues['Venue Category'].replace("Neighborhood", "Neighborhood Area") 

#### had to replace "Neighborhood" with another title, as it's a designated neighborhood of interest in Foursquare

## 3. Analyze Each Neighborhood

In [39]:
# one hot encoding
Tor_onehot = pd.get_dummies(Tor_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Tor_onehot['Neighborhood'] = Tor_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Tor_onehot.columns[-1]] + list(Tor_onehot.columns[:-1])
Tor_onehot = Tor_onehot[fixed_columns]

Tor_onehot.head()

Neighborhood  Accessories Store  Airport  Airport Food Court  \
0                  Parkwoods                  0        0                   0   
1                  Parkwoods                  0        0                   0   
2           Victoria Village                  0        0                   0   
3           Victoria Village                  0        0                   0   
4  Regent Park, Harbourfront                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    1   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     1    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0      0                0         0           0                   0   
1      0                0         0           0                   0   
2      0                0         0           0                   0   
3      0                0         0           0                   0   
4      0                0         0           0                   0   

   Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0          0       0              0          0         0   
1          0       0              0          0         0   
2          0       0              0          0         0   
3          0       0              0          0         0   
4          0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0                0         0             0

In [40]:
j_grouped=j.groupby(j['Neighbourhood']).mean().reset_index()
j_grouped

Neighbourhood   Latitude  Longitude
0                                           Agincourt  43.636847 -79.428191
1                              Alderwood, Long Branch  43.667856 -79.532242
2     Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259
3                                     Bayview Village  43.705369 -79.349372
4                   Bedford Park, Lawrence Manor East  43.651494 -79.375418
5                                         Berczy Park  43.757490 -79.374714
6                         Birch Cliff, Cliffside West  43.650571 -79.384568
7        Brockton, Parkdale Village, Exhibition Place  43.659526 -79.340923
8   Business reply mail Processing Centre, South C...  43.688905 -79.554724
9   CN Tower, King and Spadina, Railway Lands, Har...  43.662744 -79.321558
10                                Caledonia-Fairbanks  43.789053 -79.408493
11              Canada Post Gateway Processing Centre  43.669005 -79.442259
12                                          Cedarbrae  43.745906 -79.352188
13                                 Central Bay Street  43.782736 -79.442259
14                                           Christie  43.753259 -79.329656
15                               Church and Wellesley  43.696319 -79.532242
16            Clarks Corners, Tam O'Shanter, Sullivan  43.661608 -79.464763
17    Cliffside, Cliffcrest, Scarborough Village West  43.667967 -79.367675
18                     Commerce Court, Victoria Hotel  43.689574 -79.383160
19                                         Davisville  43.713756 -79.490074
20                                   Davisville North  43.653206 -79.400049
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn  43.644771 -79.373306
22                                          Don Mills  43.746375 -79.294440
23  Dorset Park, Wexford Heights, Scarborough Town...  43.672710 -79.405678
24                                          Downsview  43.675542 -79.371499
25                       Dufferin, Dovercourt Village  43.739015 -79.506944
26      East Toronto, Broadview North (Old East York)  43.706397 -79.309937
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.803762 -79.363452
28                       Fairview, Henry Farm, Oriole  43.761631 -79.520999
29             First Canadian Place, Underground city  43.724766 -79.532242
30    Forest Hill North & West, Forest Hill Road Park  43.628947 -79.394420
31                           Garden District, Ryerson  43.692657 -79.264848
32                                          Glencairn  43.757410 -79.273304
33                    Golden Mile, Clairlea, Oakridge  43.728020 -79.388790
34                  Guildwood, Morningside, West Hill  43.778517 -79.346556
35  Harbourfront East, Union Station, Toronto Islands  43.695344 -79.318389
36                      High Park, The Junction South  43.646435 -79.374846
37                                  Hillcrest Village  43.725900 -79.340923
38                                      Humber Summit  43.679563 -79.377529
39                                   Humberlea, Emery  43.657952 -79.387383
40                                 Humewood-Cedarvale  43.836125 -79.205636
41                     India Bazaar, The Beaches West  43.704324 -79.388790
42            Islington Avenue, Humber Valley Village  43.744734 -79.239476
43        Kennedy Park, Ionview, East Birchmount Park  43.709060 -79.363452
44          Kensington Market, Chinatown, Grange Park  43.651571 -79.484450
45  Kingsview Village, St. Phillips, Martin Grove ...  43.647927 -79.419750
46                   Lawrence Manor, Lawrence Heights  43.770992 -79.216917
47                                      Lawrence Park  43.648198 -79.379817
48                                            Leaside  43.752758 -79.400049
49                           Little Portugal, Trinity  43.676357 -79.293031
50                                     Malvern, Rouge  43.727929 -79.262029
51  Milliken, Agincourt North, Steeles East, L'Amo...  43.662301 -79.389494
52  Mimico N

In [41]:

Tor_grouped = Tor_onehot.groupby(Tor_onehot['Neighborhood']).mean().reset_index()
Tor_grouped

Neighborhood  Accessories Store  \
0                                           Agincourt           0.000000   
1     Bathurst Manor, Wilson Heights, Downsview North           0.000000   
2                                     Bayview Village           0.000000   
3                   Bedford Park, Lawrence Manor East           0.000000   
4                         Birch Cliff, Cliffside West           0.000000   
5        Brockton, Parkdale Village, Exhibition Place           0.000000   
6   Business reply mail Processing Centre, South C...           0.000000   
7   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
8                                 Caledonia-Fairbanks           0.000000   
9               Canada Post Gateway Processing Centre           0.000000   
10                                          Cedarbrae           0.000000   
11                                 Central Bay Street           0.000000   
12                                           Christie           0.000000   
13                               Church and Wellesley           0.000000   
14            Clarks Corners, Tam O'Shanter, Sullivan           0.000000   
15    Cliffside, Cliffcrest, Scarborough Village West           0.000000   
16                     Commerce Court, Victoria Hotel           0.000000   
17                                         Davisville           0.000000   
18                                   Davisville North           0.000000   
19   Del Ray, Mount Dennis, Keelsdale and Silverthorn           0.000000   
20                                          Don Mills           0.000000   
21  Dorset Park, Wexford Heights, Scarborough Town...           0.000000   
22                                          Downsview           0.000000   
23                       Dufferin, Dovercourt Village           0.000000   
24      East Toronto, Broadview North (Old East York)           0.000000   
25  Eringate, Bloordale Gardens, Old Burnhamthorpe...           0.000000   
26                       Fairview, Henry Farm, Oriole           0.000000   
27             First Canadian Place, Underground city           0.000000   
28    Forest Hill North & West, Forest Hill Road Park           0.000000   
29                           Garden District, Ryerson           0.000000   
30                                          Glencairn           0.000000   
31                    Golden Mile, Clairlea, Oakridge           0.000000   
32                  Guildwood, Morningside, West Hill           0.000000   
33  Harbourfront East, Union Station, Toronto Islands           0.000000   
34                      High Park, The Junction South           0.000000   
35                                  Hillcrest Village           0.000000   
36                                      Humber Summit           0.000000   
37                                   Humberlea, Emery           0.000000   
38                     India Bazaar, The Beaches West           0.000000   
39            Islington Avenue, Humber Valley Village           0.000000   
40        Kennedy Park, Ionview, East Birchmount Park           0.000000   
41          Kensington Market, Chinatown, Grange Park           0.000000   
42  Kingsview Village, St. Phillips, Martin Grove ...           0.000000   
43                   Lawrence Manor, Lawrence Heights           0.000000   
44                                      Lawrence Park           0.000000   
45                                            Leaside           0.000000   
46                           Little Portugal, Trinity           0.000000   
47                                     Malvern, Rouge           0.000000   
48  Milliken, Agincourt North, Steeles East, L'Amo...           0.000000   
49  Mimico NW, The Queensway West, South of Bloor,...           0.000000   
50                        Moore Park, Summerhill East           0.000000   
51       New Toronto, Mimico South, Humber Bay Shores           0.000000   
52           

In [42]:
Tor_grouped.shape

(96, 270)

In [43]:
num_top_venues = 10

for hood in Tor_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Tor_grouped[Tor_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                    venue  freq
0                    Café  0.12
1          Breakfast Spot  0.08
2               Nightclub  0.08
3                  Bakery  0.08
4             Coffee Shop  0.08
5       Convenience Store  0.04
6      Italian Restaurant  0.04
7           Burrito Place  0.04
8                 Stadium  0.04
9  Furniture / Home Store  0.04


----Bathurst Manor, Wilson Heights, Downsview North----
                 venue  freq
0                 Bank  0.09
1          Coffee Shop  0.09
2            Gift Shop  0.04
3          Bridal Shop  0.04
4    Mobile Phone Shop  0.04
5        Shopping Mall  0.04
6  Fried Chicken Joint  0.04
7   Frozen Yogurt Shop  0.04
8       Sandwich Place  0.04
9          Supermarket  0.04


----Bayview Village----
               venue  freq
0  Indian Restaurant  0.11
1     Sandwich Place  0.11
2        Yoga Studio  0.05
3       Burger Joint  0.05
4        Pizza Place  0.05
5        Supermarket  0.05
6     Discount Store  0.05
7          

### Convert into a pandas Dataframe

In [44]:
# Function to sort venues in decending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [45]:
Tor_grouped.shape

(96, 270)

In [46]:
# get dbs of top 10 venues per neighborhood

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Tor_grouped['Neighborhood']

for ind in np.arange(Tor_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Tor_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                  Café   
1  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
2                                  Bayview Village     Indian Restaurant   
3                Bedford Park, Lawrence Manor East           Coffee Shop   
4                      Birch Cliff, Cliffside West           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot             Nightclub                Bakery   
1           Coffee Shop           Bridal Shop                 Diner   
2        Sandwich Place           Yoga Studio                  Bank   
3                  Café             Gastropub          Cocktail Bar   
4                  Café            Restaurant        Clothing Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Coffee Shop          Climbing Gym         Burrito Place   
1           Supermarket      Sushi Restaurant     Mobile Phone Shop   
2                   Gym           Gas Station  Fast Food Restaurant   
3   American Restaurant    Italian Restaurant                Bakery   
4                   Gym                Bakery       Thai Restaurant   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Stadium    Italian Restaurant             Restaurant  
1  Middle Eastern Restaurant        Ice Cream Shop     Chinese Restaurant  
2             Discount Store                  Park               Pharmacy  
3             Cosmetics Shop              Creperie       Department Store  
4              Deli / Bodega                 Hotel    American Restaurant

In [47]:
neighborhoods_venues_sorted.shape

(96, 11)

### Clustering Neighborhoods using k-means 

In [48]:
# set number of clusters
kclusters = 5

Tor_grouped_clustering = Tor_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Tor_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([2, 2, 2, 2, 2, 2, 0, 0, 2, 2], dtype=int32)

In [49]:
neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                  Café   
1  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
2                                  Bayview Village     Indian Restaurant   
3                Bedford Park, Lawrence Manor East           Coffee Shop   
4                      Birch Cliff, Cliffside West           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot             Nightclub                Bakery   
1           Coffee Shop           Bridal Shop                 Diner   
2        Sandwich Place           Yoga Studio                  Bank   
3                  Café             Gastropub          Cocktail Bar   
4                  Café            Restaurant        Clothing Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Coffee Shop          Climbing Gym         Burrito Place   
1           Supermarket      Sushi Restaurant     Mobile Phone Shop   
2                   Gym           Gas Station  Fast Food Restaurant   
3   American Restaurant    Italian Restaurant                Bakery   
4                   Gym                Bakery       Thai Restaurant   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Stadium    Italian Restaurant             Restaurant  
1  Middle Eastern Restaurant        Ice Cream Shop     Chinese Restaurant  
2             Discount Store                  Park               Pharmacy  
3             Cosmetics Shop              Creperie       Department Store  
4              Deli / Bodega                 Hotel    American Restaurant

In [50]:
j_grouped = j_grouped.rename(columns={'Neighbourhood': 'Neighborhood'})
print(j_grouped.shape)
j_grouped.head()

(99, 3)


Neighborhood   Latitude  Longitude
0                                        Agincourt  43.636847 -79.428191
1                           Alderwood, Long Branch  43.667856 -79.532242
2  Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259
3                                  Bayview Village  43.705369 -79.349372
4                Bedford Park, Lawrence Manor East  43.651494 -79.375418

In [51]:
print(neighborhoods_venues_sorted.shape)
neighborhoods_venues_sorted.head()

(96, 11)


Neighborhood 1st Most Common Venue  \
0                                        Agincourt                  Café   
1  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
2                                  Bayview Village     Indian Restaurant   
3                Bedford Park, Lawrence Manor East           Coffee Shop   
4                      Birch Cliff, Cliffside West           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot             Nightclub                Bakery   
1           Coffee Shop           Bridal Shop                 Diner   
2        Sandwich Place           Yoga Studio                  Bank   
3                  Café             Gastropub          Cocktail Bar   
4                  Café            Restaurant        Clothing Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0           Coffee Shop          Climbing Gym         Burrito Place   
1           Supermarket      Sushi Restaurant     Mobile Phone Shop   
2                   Gym           Gas Station  Fast Food Restaurant   
3   American Restaurant    Italian Restaurant                Bakery   
4                   Gym                Bakery       Thai Restaurant   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Stadium    Italian Restaurant             Restaurant  
1  Middle Eastern Restaurant        Ice Cream Shop     Chinese Restaurant  
2             Discount Store                  Park               Pharmacy  
3             Cosmetics Shop              Creperie       Department Store  
4              Deli / Bodega                 Hotel    American Restaurant

In [52]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

In [53]:
neighborhoods_venues_sorted.head()

Cluster Labels                                     Neighborhood  \
0               2                                        Agincourt   
1               2  Bathurst Manor, Wilson Heights, Downsview North   
2               2                                  Bayview Village   
3               2                Bedford Park, Lawrence Manor East   
4               2                      Birch Cliff, Cliffside West   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                  Café        Breakfast Spot             Nightclub   
1                  Bank           Coffee Shop           Bridal Shop   
2     Indian Restaurant        Sandwich Place           Yoga Studio   
3           Coffee Shop                  Café             Gastropub   
4           Coffee Shop                  Café            Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                Bakery           Coffee Shop          Climbing Gym   
1                 Diner           Supermarket      Sushi Restaurant   
2                  Bank                   Gym           Gas Station   
3          Cocktail Bar   American Restaurant    Italian Restaurant   
4        Clothing Store                   Gym                Bakery   

  7th Most Common Venue      8th Most Common Venue 9th Most Common Venue  \
0         Burrito Place                    Stadium    Italian Restaurant   
1     Mobile Phone Shop  Middle Eastern Restaurant        Ice Cream Shop   
2  Fast Food Restaurant             Discount Store                  Park   
3                Bakery             Cosmetics Shop              Creperie   
4       Thai Restaurant              Deli / Bodega                 Hotel   

  10th Most Common Venue  
0             Restaurant  
1     Chinese Restaurant  
2               Pharmacy  
3       Department Store  
4    American Restaurant

In [54]:
# join
Tor_merged = j_grouped

# merge Tor_grouped with j to add latitude/longitude for each neighborhood

#Tor_merged = Tor_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood',how='left')
Tor_merged = Tor_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Tor_merged.head() # check the last columns!

Neighborhood   Latitude  Longitude  \
0                                        Agincourt  43.636847 -79.428191   
1                           Alderwood, Long Branch  43.667856 -79.532242   
2  Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259   
3                                  Bayview Village  43.705369 -79.349372   
4                Bedford Park, Lawrence Manor East  43.651494 -79.375418   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0             2.0                  Café        Breakfast Spot   
1             NaN                   NaN                   NaN   
2             2.0                  Bank           Coffee Shop   
3             2.0     Indian Restaurant        Sandwich Place   
4             2.0           Coffee Shop                  Café   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0             Nightclub                Bakery           Coffee Shop   
1                   NaN                   NaN                   NaN   
2           Bridal Shop                 Diner           Supermarket   
3           Yoga Studio                  Bank                   Gym   
4             Gastropub          Cocktail Bar   American Restaurant   

  6th Most Common Venue 7th Most Common Venue      8th Most Common Venue  \
0          Climbing Gym         Burrito Place                    Stadium   
1                   NaN                   NaN                        NaN   
2      Sushi Restaurant     Mobile Phone Shop  Middle Eastern Restaurant   
3           Gas Station  Fast Food Restaurant             Discount Store   
4    Italian Restaurant                Bakery             Cosmetics Shop   

  9th Most Common Venue 10th Most Common Venue  
0    Italian Restaurant             Restaurant  
1                   NaN                    NaN  
2        Ice Cream Shop     Chinese Restaurant  
3                  Park               Pharmacy  
4              Creperie       Department Store

In [55]:
Tor_merged.columns

Index(['Neighborhood', 'Latitude', 'Longitude', 'Cluster Labels',
       '1st Most Common Venue', '2nd Most Common Venue',
       '3rd Most Common Venue', '4th Most Common Venue',
       '5th Most Common Venue', '6th Most Common Venue',
       '7th Most Common Venue', '8th Most Common Venue',
       '9th Most Common Venue', '10th Most Common Venue'],
      dtype='object')

In [56]:
Tor_merged=Tor_merged.dropna()
Tor_merged.reset_index()
Tor_merged.head()

Neighborhood   Latitude  Longitude  \
0                                        Agincourt  43.636847 -79.428191   
2  Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259   
3                                  Bayview Village  43.705369 -79.349372   
4                Bedford Park, Lawrence Manor East  43.651494 -79.375418   
6                      Birch Cliff, Cliffside West  43.650571 -79.384568   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0             2.0                  Café        Breakfast Spot   
2             2.0                  Bank           Coffee Shop   
3             2.0     Indian Restaurant        Sandwich Place   
4             2.0           Coffee Shop                  Café   
6             2.0           Coffee Shop                  Café   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0             Nightclub                Bakery           Coffee Shop   
2           Bridal Shop                 Diner           Supermarket   
3           Yoga Studio                  Bank                   Gym   
4             Gastropub          Cocktail Bar   American Restaurant   
6            Restaurant        Clothing Store                   Gym   

  6th Most Common Venue 7th Most Common Venue      8th Most Common Venue  \
0          Climbing Gym         Burrito Place                    Stadium   
2      Sushi Restaurant     Mobile Phone Shop  Middle Eastern Restaurant   
3           Gas Station  Fast Food Restaurant             Discount Store   
4    Italian Restaurant                Bakery             Cosmetics Shop   
6                Bakery       Thai Restaurant              Deli / Bodega   

  9th Most Common Venue 10th Most Common Venue  
0    Italian Restaurant             Restaurant  
2        Ice Cream Shop     Chinese Restaurant  
3                  Park               Pharmacy  
4              Creperie       Department Store  
6                 Hotel    American Restaurant

In [57]:
Tor_merged.head()

Neighborhood   Latitude  Longitude  \
0                                        Agincourt  43.636847 -79.428191   
2  Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259   
3                                  Bayview Village  43.705369 -79.349372   
4                Bedford Park, Lawrence Manor East  43.651494 -79.375418   
6                      Birch Cliff, Cliffside West  43.650571 -79.384568   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0             2.0                  Café        Breakfast Spot   
2             2.0                  Bank           Coffee Shop   
3             2.0     Indian Restaurant        Sandwich Place   
4             2.0           Coffee Shop                  Café   
6             2.0           Coffee Shop                  Café   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0             Nightclub                Bakery           Coffee Shop   
2           Bridal Shop                 Diner           Supermarket   
3           Yoga Studio                  Bank                   Gym   
4             Gastropub          Cocktail Bar   American Restaurant   
6            Restaurant        Clothing Store                   Gym   

  6th Most Common Venue 7th Most Common Venue      8th Most Common Venue  \
0          Climbing Gym         Burrito Place                    Stadium   
2      Sushi Restaurant     Mobile Phone Shop  Middle Eastern Restaurant   
3           Gas Station  Fast Food Restaurant             Discount Store   
4    Italian Restaurant                Bakery             Cosmetics Shop   
6                Bakery       Thai Restaurant              Deli / Bodega   

  9th Most Common Venue 10th Most Common Venue  
0    Italian Restaurant             Restaurant  
2        Ice Cream Shop     Chinese Restaurant  
3                  Park               Pharmacy  
4              Creperie       Department Store  
6                 Hotel    American Restaurant

In [58]:
Tor_merged.dtypes
Tor_merged['Cluster Labels']=Tor_merged['Cluster Labels'].astype('int64') 
#df['cluster_labels']=df['cluster_labels'].astype('int64') 
Tor_merged

Neighborhood   Latitude  Longitude  \
0                                           Agincourt  43.636847 -79.428191   
2     Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259   
3                                     Bayview Village  43.705369 -79.349372   
4                   Bedford Park, Lawrence Manor East  43.651494 -79.375418   
6                         Birch Cliff, Cliffside West  43.650571 -79.384568   
7        Brockton, Parkdale Village, Exhibition Place  43.659526 -79.340923   
8   Business reply mail Processing Centre, South C...  43.688905 -79.554724   
9   CN Tower, King and Spadina, Railway Lands, Har...  43.662744 -79.321558   
10                                Caledonia-Fairbanks  43.789053 -79.408493   
11              Canada Post Gateway Processing Centre  43.669005 -79.442259   
12                                          Cedarbrae  43.745906 -79.352188   
13                                 Central Bay Street  43.782736 -79.442259   
14                                           Christie  43.753259 -79.329656   
15                               Church and Wellesley  43.696319 -79.532242   
16            Clarks Corners, Tam O'Shanter, Sullivan  43.661608 -79.464763   
17    Cliffside, Cliffcrest, Scarborough Village West  43.667967 -79.367675   
18                     Commerce Court, Victoria Hotel  43.689574 -79.383160   
19                                         Davisville  43.713756 -79.490074   
20                                   Davisville North  43.653206 -79.400049   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn  43.644771 -79.373306   
22                                          Don Mills  43.746375 -79.294440   
23  Dorset Park, Wexford Heights, Scarborough Town...  43.672710 -79.405678   
24                                          Downsview  43.675542 -79.371499   
25                       Dufferin, Dovercourt Village  43.739015 -79.506944   
26      East Toronto, Broadview North (Old East York)  43.706397 -79.309937   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.803762 -79.363452   
28                       Fairview, Henry Farm, Oriole  43.761631 -79.520999   
29             First Canadian Place, Underground city  43.724766 -79.532242   
30    Forest Hill North & West, Forest Hill Road Park  43.628947 -79.394420   
31                           Garden District, Ryerson  43.692657 -79.264848   
32                                          Glencairn  43.757410 -79.273304   
33                    Golden Mile, Clairlea, Oakridge  43.728020 -79.388790   
34                  Guildwood, Morningside, West Hill  43.778517 -79.346556   
35  Harbourfront East, Union Station, Toronto Islands  43.695344 -79.318389   
36                      High Park, The Junction South  43.646435 -79.374846   
37                                  Hillcrest Village  43.725900 -79.340923   
38                                      Humber Summit  43.679563 -79.377529   
39                                   Humberlea, Emery  43.657952 -79.387383   
41                     India Bazaar, The Beaches West  43.704324 -79.388790   
42            Islington Avenue, Humber Valley Village  43.744734 -79.239476   
43        Kennedy Park, Ionview, East Birchmount Park  43.709060 -79.363452   
44          Kensington Market, Chinatown, Grange Park  43.651571 -79.484450   
45  Kingsview Village, St. Phillips, Martin Grove ...  43.647927 -79.419750   
46                   Lawrence Manor, Lawrence Heights  43.770992 -79.216917   
47                                      Lawrence Park  43.648198 -79.379817   
48                                            Leaside  43.752758 -79.400049   
49                           Little Portugal, Trinity  43.676357 -79.293031   
50                                     Malvern, Rouge  43.727929 -79.262029   
51  Milliken, Agincourt North, Steeles East, L'Amo...  43.662301 -79.389494   
52  Mimico NW, The Queensway West, South of Bloor,...  43.706748 -79.594054   
53          

In [59]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Tor_merged['Latitude'], Tor_merged['Longitude'], Tor_merged['Neighborhood'], Tor_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Final clusters for all of the neighborhoods:  most clustered into 2 of the 5 clusters -- Cluster 0 was more recreational/sports and cluster 2 was more cafes and restaurants.  The remaining clusteres were negligible.